### DISCUSSION TITLE
Feature Request: Assertion / Error / Invariant

### DISCUSSION THREAD

**Discussion by danfrankj** [duckdb/duckdb#6775] [Ideas]
Why do you want this feature?

Sometimes you want to force SQL execution to stop with a helpful error. eg. if there's some unexpected quality of the data 

Example
```
SELECT 42 / denom
FROM (
  -- some input table which may or may not have 0s
  SELECT 0 AS denom
)
```

This will output a NULL or NAN which I have to handle later

Instead, I'd like to be able to do something like 
```
SELECT 42 / denom 
FROM ( 
  SELECT denom, INVARIANT(denom != 0, 'denominator must not be zero')
  FROM ( 
    -- some input table which may or may not have 0s
    SELECT 0 AS denom
  )  
) 
```

Which would give me a helpful error message and make it so I don't have to handle the somewhat unpredictable behavior of what happens if 1/0.

**Comment by ozdemircs** [comment]
Hi, thanks for adding this idea, I will try take this as a first time contributor.

To verify the assumptions,

INVARIANT function will return true when the condition is true.
INVARIANT function will fail the execution with an error message given as a second argument if the condition is false.
INVARIANT function will return NULL if the condition is NULL.

Does that sound good?

**Reply by ozdemircs** [reply]
Actually, I have written a small scalar function for what I had in my mind, but I think it does not work as I expected. When using the invariant at the top level, it gives an error message, while if we use the invariant function as part of the FROM statement, the error is ignored  at the top level. Although in tests, both queries fail as expected.

```
D SELECT denom, INVARIANT(denom != 0, 'denominator must not be zero')
> FROM ( 
> SELECT 1 AS denom
> UNION ALL
> SELECT 0 AS denom
> UNION ALL
> SELECT 1 AS denom
> );
Error: Invalid Error: denominator must not be zero
D SELECT 42/denom FROM (
> SELECT denom, INVARIANT(denom != 0, 'denominator must not be zero')
> FROM ( 
> SELECT 1 AS denom
> UNION ALL
> SELECT 0 AS denom
> UNION ALL
> SELECT 1 AS denom
> ));
┌──────────────┐
│ (42 / denom) │
│    int32     │
├──────────────┤
│           42 │
│              │
│           42 │
└──────────────┘
D 
```

**Reply by Tishj** [reply]
Actually looking at this again, the reason this happens is because of projection pushdown.
The results of INVARIANT are not used, so it's stripped from the optimized plan, which we can see when we add EXPLAIN to your query:
```
physical_plan   
┌───────────────────────────┐                                                          
│         PROJECTION        │                                                          
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                                                          
│        (42 / denom)       │                                                          
└─────────────┬─────────────┘                                                                                       
┌─────────────┴─────────────┐                                                          
│           UNION           ├───────────────────────────────────────────┐              
└─────────────┬─────────────┘                                           │                                           
┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐
│           UNION           │                             │         PROJECTION        │
│                           ├──────────────┐              │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│                           │              │              │           denom           │
└─────────────┬─────────────┘              │              └─────────────┬─────────────┘                             
┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         PROJECTION        ││         PROJECTION        ││         DUMMY_SCAN        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │
│           denom           ││           denom           ││                           │
└─────────────┬─────────────┘└─────────────┬─────────────┘└───────────────────────────┘                             
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                             
│         DUMMY_SCAN        ││         DUMMY_SCAN        │                             
└───────────────────────────┘└───────────────────────────┘ 
```

When we disable the `RemoveUnusedColumns` optimizer pass, we get the expected behavior.
(`optimizer.cpp`:122)
```c++
	// removes unused columns
	RunOptimizer(OptimizerType::UNUSED_COLUMNS, [&]() {
		RemoveUnusedColumns unused(binder, context, true);
		unused.VisitOperator(*plan);
	});
```
You likely need to make certain optimizer rules (starting with RemoveUnusedColumns) aware of the `INVARIANT` function and don't let them act on columns created by it

**Reply by ozdemircs** [reply]
Thanks @Tishj I really appreciate your help in here and I think this explanation totally makes sense to me, only one question I have, why does the test case return an error for the same query while the query through the console client does not? 

I would assume the execution graph for the test case perhaps is not optimized, could it be the reason?

For the solution suggestions,

I see three options,

1. Trying to make the optimizer aware of the invariant function so that it is not stripped as you suggested.
2. Evolve it similar to the ERROR function @danfrankj points out. BigQuery also does not have any guarantee for it to be evaluated, looking at the examples, I think, it is likely for similar reasons.
3. Evolve it to return some expression such that user might be using for the further.

```
// name can be decided later, this returns the 3rd parameter if the first condition is not correct or throws an error.
IF_INVARIANT(denom != 0, 'error message', denom) as denom
```

I think, first option is a bit less flexible for two reasons, first, besides the unused, other rules can also discard the INVARIANT function, also it is possible the changes we made in the optimizer code can complicate the optimizer code without a strong use case.

Second option, although it is not ideal, having a similar function in BigQuery is a good indicator what we do is not uncommon, we could also evolve to ERROR function as @danfrankj points.

In the third option, users can use the output of the function for the parent query so that the function is not stripped, but there is still no guarantee it will be stripped for different reasons.

What I suggest is to go with the ERROR function in option 2, and document (not sure if these functions have documentations) the explain statement to check by the user if the ERROR will be discarded or not.

Please let me know what you think.

**Reply by Tishj** [reply]
I think 3 sounds like a good idea, especially with the addition of the new `.` syntax for chaining operations
`function(input.ASSERT(x -> x != NULL))` looks like nice syntax
(it would essentially be shorthand for: `function(ASSERT(input, x -> x != NULL))`, see https://github.com/duckdb/duckdb/pull/6725)

**Comment by danfrankj** [comment]
It might actually make more sense to do a simpler "ERROR()" function. INVARIANT can easily be built on top of that. But as mentioned, the hard part is making sure ERROR() is conditionally evaluated 

Example from bigquery https://cloud.google.com/bigquery/docs/reference/standard-sql/debugging_functions

**Comment by Tishj** [comment]
Relevant branch of the ozdemircs:
https://github.com/duckdb/duckdb/compare/master...ozdemircs:duckdb:invariants_scalar_function

**Reply by Tishj** [reply]
I don't spot any glaring mistakes, I'm not that familiar with our UnaryExecutor and BinaryExecutor inner workings though.

And this could also be due to default null handling, now that I think about it. 🤔 

Do scalar functions have a null handling option? This should likely be SPECIAL_HANDLING if so

Edit: they do, and I think it's relevant here

**Comment by Mytherin** [comment]
The `error` function has been added sometime ago, see the [documentation](https://duckdb.org/docs/sql/functions/utility.html#errormessage)

```sql
select error('my error message');
```
